core/lib/dal/src/models/storage_event.rs

```diff
@@ -19,10 +20,49 @@ pub struct StorageWeb3Log {
     pub tx_index_in_block: i32,
     pub event_index_in_block: i32,
     pub event_index_in_tx: i32,
+    pub event_index_in_block_without_eth_transfer: Option<i32>,
+    pub event_index_in_tx_without_eth_transfer: Option<i32>,
 }
 
-impl From<StorageWeb3Log> for Log {
-    fn from(log: StorageWeb3Log) -> Log {
+impl StorageWeb3Log {
+    pub fn maybe_into_extended_storage_log(
+        self,
+        api_eth_transfer_events: ApiEthTransferEvents,
+    ) -> Option<ExtendedStorageWeb3Log> {
+        if api_eth_transfer_events == ApiEthTransferEvents::Disabled
+            && self.address == L2_ETH_TOKEN_ADDRESS.as_bytes()
+            && self.topic1 == TRANSFER_EVENT_TOPIC.as_bytes()
+        {
+            None
+        } else {
+            Some(ExtendedStorageWeb3Log(
+                self.clone(),
+                api_eth_transfer_events,
+            ))
+        }
+    }
+}
+
+pub struct ExtendedStorageWeb3Log(pub StorageWeb3Log, pub ApiEthTransferEvents);
+
+impl From<ExtendedStorageWeb3Log> for Log {
+    fn from(log_with_transfer_events_mode: ExtendedStorageWeb3Log) -> Log {
+        let log = log_with_transfer_events_mode.0;
+        let api_eth_transfer_events = log_with_transfer_events_mode.1;
+
+        let event_index_in_block = match api_eth_transfer_events {
+            ApiEthTransferEvents::Enabled => log.event_index_in_block,
+            ApiEthTransferEvents::Disabled => {
+                log.event_index_in_block_without_eth_transfer.unwrap_or(0)
```

Comment:
should it panic? I mean if it returns 0 in API it might lead to worse consequences